Fast-path conversions between Temporal types

ptomato opened the pull request:
Not all the changes have test262 tests yet, but writing them was taking so long that I decided to go ahead and put this up for review anyway. I'll add the undone ones to #1398.

Closes: #1428

ljharb commented on spec/calendar.html:
```diff
@@ -237,9 +237,15 @@ <h1>CalendarInLeapYear ( _calendar_, _dateLike_ )</h1>
-          1. Return _temporalCalendarLike_.
+          1. If _temporalCalendarLike_ has an [[InitializedTemporalDate]], [[InitializedTemporalDateTime]], [[InitializedTemporalMonthDay]], [[InitializedTemporalTime]], [[InitializedTemporalYearMonth]], or [[InitializedTemporalZonedDateTime]] internal slot, then
+            1. Return _temporalCalendarLike_.[[Calendar]].
+          1. If ? HasProperty(_item_, *"calendar"*) is *false*, return _item_.
+          1. Set _item_ to ? Get(_item_, *"calendar"*).
+          1. If Type(_item_) is Object and ? HasProperty(_item_, *"calendar"*) is *false*, return _item_.
```
can you remind me why this is checking `temporalCalendarLike.calendar.calendar`?

ptomato replied:
The consensus is documented in https://github.com/tc39/proposal-temporal/issues/925#issuecomment-709544520 — tl;dr this is to allow `{ calendar: 'gregory', year: 2021, month: 4, day: 13 }` as an input, but not `{ calendar: { calendar: { calendar: { ... } } } }`.

ljharb commented on spec/calendar.html:
```diff
@@ -237,9 +237,15 @@ <h1>CalendarInLeapYear ( _calendar_, _dateLike_ )</h1>
-          1. Return _temporalCalendarLike_.
+          1. If _temporalCalendarLike_ has an [[InitializedTemporalDate]], [[InitializedTemporalDateTime]], [[InitializedTemporalMonthDay]], [[InitializedTemporalTime]], [[InitializedTemporalYearMonth]], or [[InitializedTemporalZonedDateTime]] internal slot, then
+            1. Return _temporalCalendarLike_.[[Calendar]].
+          1. If ? HasProperty(_item_, *"calendar"*) is *false*, return _item_.
+          1. Set _item_ to ? Get(_item_, *"calendar"*).
```
```suggestion
          1. If ? HasProperty(_temporalCalendarLike_, *"calendar"*) is *false*, return _item_.
          1. Set _item_ to ? Get(_temporalCalendarLike_, *"calendar"*).
```

ljharb commented on spec/plaintime.html:
```diff
@@ -610,11 +612,14 @@ <h1>ToTemporalTime ( _item_ [ , _overflow_ ] )</h1>
+          1. If _item_ has an [[InitializedTemporalZonedDateTime]] internal slot, then
+            1. Let _instant_ be ? CreateTemporalInstant(_item_.[[Nanoseconds]]).
+            1. Set _item_ to ? BuiltinTimeZoneGetPlainDateTimeFor(_item_.[[TimeZone]], _instant_, _item_.[[Calendar]]).
+          1. If _item_ has an [[InitializedTemporalDateTime]] internal slot, then
```
```suggestion
          1. Else if _item_ has an [[InitializedTemporalDateTime]] internal slot, then
```
since these paths should be mutually exclusive?

ptomato replied:
This particular path is not mutually exclusive (in lines 615–617 _item_ is modified from a ZonedDateTime into a PlainDateTime and subsequently "falls through" into the PlainDateTime clause) but reading it back, maybe it's better to be more explicit and not write it this way, because I can see where it could be confusing.

ljharb replied:
ah, i see what you mean. i definitely find it confusing to reuse variable names, both in the spec and in actual code :-)

ptomato replied:
Assuming that you're OK with "If" and not "Else if" if the previous clause Returns, I've fixed this up everywhere it occurs.

cjtenny reviewed (commented):
Part 1: Reviewed everything but the tests. Wanted to offer the feedback sooner since I doubt I'll have changes to suggest for the tests :)

ptomato commented on polyfill/lib/calendar.mjs:
```diff
@@ -94,17 +94,17 @@ export class Calendar {
     if (!ES.IsTemporalCalendar(this)) throw new TypeError('invalid receiver');
     if (ES.IsTemporalMonthDay(date)) throw new TypeError('use monthCode on PlainMonthDay instead');
-    if (!HasSlot(date, ISO_YEAR)) date = ES.ToTemporalDate(date);
+    if (!HasSlot(date, ISO_MONTH)) date = ES.ToTemporalDate(date);
```
Oh, hmm, I forgot when writing this that we renamed the [[ReferenceISOYear]] and [[ReferenceISODay]] slots into just [[ISOYear]] and [[ISODay]]. This change has no effect, then.

My intention was to make sure, now that we have a fast path from PlainDateTime/ZonedDateTime to PlainDate, that anything other than PlainDate, PlainYearMonth, and PlainMonthDay is passed through ToTemporalDate. But even as I had written it in the PR, that wouldn't have been the effect; sorry, it looks like I was rushing too much to get this done yesterday evening. I will try a different approach.

ptomato commented on spec/calendar.html:
```diff
@@ -237,9 +237,15 @@ <h1>CalendarInLeapYear ( _calendar_, _dateLike_ )</h1>
+          1. If Type(_item_) is Object and ? HasProperty(_item_, *"calendar"*) is *false*, return _item_.
         1. Let _identifier_ be ? ToString(_temporalCalendarLike_).
-        1. Return ? CalendarFrom(_identifier_, %Temporal.Calendar%).
+        1. If ! IsBuiltinCalendar(_identifier_) is *false*, then
```
This isn't a negation, it's an assertion that the operation doesn't throw. https://tc39.es/ecma262/#sec-returnifabrupt-shorthands

cjtenny replied:
:laughing: my bad, it was late! Other excuses!

ptomato commented on spec/timezone.html:
```diff
@@ -534,21 +534,14 @@ <h1>FormatTimeZoneOffsetString ( _offsetNanoseconds_ )</h1>
+            1. Return _temporalTimeZoneLike_.[[TimeZone]].
           1. If ? HasProperty(_item_, *"timeZone"*) is *false*, return _item_.
           1. Set _item_ to ? Get(_item_, *"timeZone"*).
           1. If Type(_item_) is Object and ? HasProperty(_item_, *"timeZone"*) is *false*, return _item_.
```
Yes, see in https://github.com/tc39/proposal-temporal/issues/925#issuecomment-709544520 — tl;dr this is to allow `{ timeZone: 'UTC', year: 2021, month: 4, day: 13 }` as an input, but not `{ timeZone: { timeZone: { timeZone: { ... } } } }`.

ptomato commented on spec/zoneddatetime.html:
```diff
@@ -1120,7 +1122,8 @@ <h1>ToTemporalZonedDateTime ( _item_ [ , _options_ ] )</h1>
           1. Let _result_ be ? ParseTemporalZonedDateTimeString(_string_).
           1. If _result_.[[TimeZoneName]] is *undefined*, throw a *RangeError* exception.
-          1. Let _timeZone_ be ? TimeZoneFrom(_result_.[[TimeZoneName]]).
+          1. If ! IsValidTimeZoneName(_result_.[[TimeZoneName]]) is *false*, throw a *RangeError* exception.
```
That is true, but I think the assertion in ParseTemporalTimeZoneString is incorrect and should be a throw instead. I'll update that.

ptomato commented:
OK, thanks for the reviews! Only the last two commits are substantially different from the previous version.

cjtenny commented on spec/intl.html:
```diff
@@ -1267,281 +1267,273 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
             1. Perform ? RequireInternalSlot(_calendar_, [[InitializedTemporalCalendar]]).
-            1. If Type(_dateOrDateTime_) is not Object or _dateOrDateTime_ does not have an [[ISOYear]] internal slot, then
-              1. Set _dateOrDateTime_ to ? ToTemporalDate(_dateOrDateTime_).
+            1. Let _temporalDate_ be ? ToTemporalDate(_temporalDateLike_).
```
Doesn't match polyfill - YearMonth doesn't need to ToDate

cjtenny commented on spec/intl.html:
```diff
@@ -1267,281 +1267,273 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
             1. Perform ? RequireInternalSlot(_calendar_, [[InitializedTemporalCalendar]]).
-            1. If Type(_dateOrDateTime_) is not Object or _dateOrDateTime_ does not have an [[ISOYear]] internal slot, then
-              1. Set _dateOrDateTime_ to ? ToTemporalDate(_dateOrDateTime_).
+            1. Let _temporalDate_ be ? ToTemporalDate(_temporalDateLike_).
```
Same as prior

cjtenny commented on spec/intl.html:
```diff
@@ -1267,281 +1267,273 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
             1. Let _calendar_ be the *this* value.
             1. Perform ? RequireInternalSlot(_calendar_, [[InitializedTemporalCalendar]]).
-            1. If Type(_dateOrDateTime_) is Object and _dateOrDateTime_ has an [[InitializedTemporalMonthDay]] internal slot, then
+            1. If Type(_temporalDateLike_) is Object and _temporalDateLike_ has an [[InitializedTemporalMonthDay]] internal slot, then
```
Doesn't match polyfill - polyfill throws the exception after performing several Get operations I think, because it falls through to ToTemporalDate. If `.month` added to prototype of PlainMonthDay is allowed behavior, then we should change this; otherwise, we can change ToTemporalDate to TypeError on receiving a PMD. WDYT?

ptomato replied:
This was a pre-existing check, that was intentional to allow implementations to throw an informative error about PlainMonthDay only having `monthCode` and not `month`. (Unfortunately I can't find the reference where this was discussed, at the moment)

cjtenny replied:
Has this been changed to match now? I went back to re-review it and it looks like it does match and throw that informative error.  In which case, resolved.

ptomato replied:
I didn't change it, but it sounds like it's correct now :smile:

cjtenny commented on spec/intl.html:
```diff
@@ -1267,281 +1267,273 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
           <emu-alg>
             1. Let _calendar_ be the *this* value.
             1. Perform ? RequireInternalSlot(_calendar_, [[InitializedTemporalCalendar]]).
+            1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]] or [[InitializedTemporalYearMonth]] internal slot, then
```
Can remove all these checks for InitializedTemporalDate for the same reason we don't do them in the polyfill - ToTemporalDate shortcuts if InitializedTemporalDate already present.

ptomato replied:
True, but I thought the intention would be clearer to the reader if I left them in. Happy to go either way.

cjtenny replied:
I'm happy either way too; worth nothing that this makes the spec more brittle to changes, but since the spec shouldn't be seeing many more changes, I'm ambivalent. Your call!

ptomato replied:
It's an editorial change if we want to change it in the future, so I'll resolve this as-is for expediency.

cjtenny reviewed (approved):
Really clever (not in a bad way - it's very clean and clear, just took a few minutes to wrap my brain around what would get called in what abstract op), well done!

I think fixes are needed to `era`, `eraYear` in polyfill, and `era`, `eraYear`, and `month` in Intl spec text. The tests look great, awesome work with the helper. Also optionally can remove redundant [[InitializedTemporalDate]] slot checks in the spec when the alternative is calling `ToTemporalDate` since `ToTemporalDate` shortcuts that path anyway, which is why the polyfill doesn't do that check.

Approved once those changes are in!

ptomato commented on spec/plaintime.html:
```diff
@@ -610,11 +612,15 @@ <h1>ToTemporalTime ( _item_ [ , _overflow_ ] )</h1>
-            1. Set _calendar_ to ? ToTemporalCalendar(_calendar_).
-            1. If ? ToString(_calendar_) is not *"iso8601"*, then
-              1. Throw a *RangeError* exception.
+          1. If _item_ has an [[InitializedTemporalZonedDateTime]] internal slot, then
```
Yes, handling Instant would require a time zone, so there's no fast path possible.

ptomato commented on polyfill/test/Instant/prototype/toZonedDateTime/plain-custom-timezone.js:
```diff
@@ -7,7 +7,10 @@ includes: [compareArray.js]
 const actual = [];
-const expected = [];
+const expected = [
+  "has timeZone.timeZone", // ToTemporalTimeZone in Instant.prototype.toZonedDateTime
```
These don't occur in the `from` calls at all, but both occur internally in `Temporal.Instant.prototype.toZonedDateTime()`. It's called once on the argument at the beginning of the method, and once when constructing the new instance of ZonedDateTime. I agree this is not ideal, and I just wondered if it might be good to change CreateZonedDateTime to _not_ call ToTemporalZonedDateTime, but instead move that out to all the callers or remove it if, as in this case, the time zone is already guaranteed to be a valid value. What do you think?

ptomato replied:
We discussed this out of band and decided that this change would be within the mandate of this consensus (before the change requested in this issue, there were zero observable Has operations, now there are two, and ideally there is only one). I'll make this change, and then land the pull request.

ptomato replied:
Of course, as soon as I go to make the change I realize that it is already this way in the spec text, but we just need to make sure it is reflected in the polyfill.

ptomato replied:
I've fixed the bug in the polyfill and will go ahead and merge this today, so we can keep the normative changes moving. I'm happy to receive comments on the polyfill change after the fact, and can fix any mistakes up anytime, since it is not normative.

ljharb commented on spec/calendar.html:
```diff
@@ -804,191 +794,200 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
         1. Assert: _calendar_.[[Identifier]] is *"iso8601"*.
-        1. Return ? ISOYear(_dateOrDateTime_).
+        1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]] or [[InitializedTemporalYearMonth]] internal slot, then
+          1. Set _temporalDateLike_ to ? ToTemporalDate(_temporalDateLike_).
```
nit:
```suggestion
          1. Return ! ISOYear(? ToTemporalDate(_temporalDateLike_)).
```
to avoid reusing the variable name?

ljharb commented on spec/calendar.html:
```diff
@@ -804,191 +794,200 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
+        1. If Type(_temporalDateLike_) is Object and _temporalDateLike_ has an [[InitializedTemporalMonthDay]] internal slot, then
+          1. Throw a *TypeError* exception.
+        1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]] or [[InitializedTemporalYearMonth]] internal slot, then
+          1. Set _temporalDateLike_ to ? ToTemporalDate(_temporalDateLike_).
+        1. Return ! ISOMonth(_temporalDateLike_).
```
same nit here:
```suggestion
          1. Return ! ISOMonth(? ToTemporalDate(_temporalDateLike_)).
        1. Return ! ISOMonth(_temporalDateLike_).
```

ljharb commented on spec/calendar.html:
```diff
@@ -804,191 +794,200 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
         1. Assert: _calendar_.[[Identifier]] is *"iso8601"*.
-        1. Return ? ISOMonthCode(_dateOrDateTime_).
+        1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]], [[InitializedTemporalMonthDay]], or [[InitializedTemporalYearMonth]] internal slot, then
+          1. Set _temporalDateLike_ to ? ToTemporalDate(_temporalDateLike_).
+        1. Return ! ISOMonthCode(_temporalDateLike_).
```
also here:
```suggestion
          1. Return ! ISOMonth(? ToTemporalDate(_temporalDateLike_)).
        1. Return ! ISOMonthCode(_temporalDateLike_).
```

ljharb commented on spec/calendar.html:
```diff
@@ -804,191 +794,200 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
         1. Assert: _calendar_.[[Identifier]] is *"iso8601"*.
-        1. Return ? ISODay(_dateOrDateTime_).
+        1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]] or [[InitializedTemporalMonthDay]] internal slot, then
+          1. Set _temporalDateLike_ to ? ToTemporalDate(_temporalDateLike_).
+        1. Return ! ISODay(_temporalDateLike_).
```
and here:
```suggestion
          1. Return ! ISODay(? ToTemporalDate(_temporalDateLike_)).
        1. Return ! ISODay(_temporalDateLike_).
```

ljharb commented on spec/calendar.html:
```diff
@@ -804,191 +794,200 @@ <h1>Temporal.Calendar.prototype.dateUntil ( _one_, _two_, _options_ )</h1>
-        1. Let _year_ be ? ISOYear(_dateOrDateTime_).
-        1. Return 𝔽(! ISODaysInYear(_year_)).
+        1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]] or [[InitializedTemporalYearMonth]] internal slot, then
+          1. Set _temporalDateLike_ to ? ToTemporalDate(_temporalDateLike_).
+        1. Return 𝔽(! ISODaysInYear(_temporalDateLike_.[[ISOYear]])).
```
altho in this case, the F() and ISODaysInYear() kind of seem cleaner as-is, so maybe all my previous suggestions should be ignored?

ptomato replied:
This was the reason why I left them all that way, though I realize the other places you marked above are not ideal. I'll leave this for now but if you change your mind I don't mind changing it either.

ljharb replied:
We’ll see what the editors say.

ljharb commented on spec/calendar.html:
```diff
@@ -1025,8 +1025,9 @@ <h1>Temporal.Calendar.prototype.inLeapYear ( _dateOrDateTime_ )</h1>
-        1. Let _year_ be ? ISOYear(_dateOrDateTime_).
-        1. Return ! IsISOLeapYear(_year_).
+        1. If Type(_temporalDateLike_) is not Object or _temporalDateLike_ does not have an [[InitializedTemporalDate]] or [[InitializedTemporalYearMonth]] internal slot, then
+          1. Set _temporalDateLike_ to ? ToTemporalDate(_temporalDateLike_).
+        1. Return ! IsISOLeapYear(_temporalDateLike_.[[ISOYear]]).
```
same here, if the other similar changes are made:
```suggestion
          1. Return ! IsISOLeapYear(? ToTemporalDate(_temporalDateLike_)).
        1. Return ! IsISOLeapYear(_temporalDateLike_.[[ISOYear]]).
```

ljharb commented on spec/plaindate.html:
```diff
@@ -374,6 +374,8 @@ <h1>Temporal.PlainDate.prototype.with ( _temporalDateLike_ [ , _options_ ] )</h1
         1. Perform ? RequireInternalSlot(_temporalDate_, [[InitializedTemporalDate]]).
         1. If Type(_temporalDateLike_) is not Object, then
           1. Throw a *TypeError* exception.
+        1. If _temporalDateLike_ has an [[InitializedTemporalDate]], [[InitializedTemporalDateTime]], [[InitializedTemporalMonthDay]], [[InitializedTemporalTime]], [[InitializedTemporalYearMonth]], or [[InitializedTemporalZonedDateTime]] internal slot, then
+          1. Throw a *TypeError* exception.
```
these three operations seem kind of like it would be valuable as some kind of `Perform ? RequireTemporalType(_temporalDate_, ~PlainDate~)` rather than laboriously listing internal slots everywhere this kind of check is needed.

ptomato replied:
I've gone with RejectTemporalCalendarType, hopefully the name makes it clear that the intention is that anything with a calendar should not make it through.

ptomato commented:
I believe that I've addressed the comments, and would like to get this landed before the plenary so I'm going to go ahead and merge it. Happy to address any further comments as they come in.

FrankYFTang commented:
@ptomato Could you help me to answer this question in https://chromium-review.googlesource.com/c/v8/v8/+/3531554/8/src/objects/js-temporal-objects.cc#6190 
?

" why the spec rejects this but allows this kind of object for monthCode? Just strikes me as odd."

This is referring to the fact that Temporal.Calendar.prototype.monthCode will accept a PlainMonthDay object but Temporal.Calendar.prototype.month will throw TypeError  exception 

The throw is added by you in this PR
See spec text here
step 4 in https://tc39.es/proposal-temporal/#sec-temporal.calendar.prototype.month
but no such step in 
https://tc39.es/proposal-temporal/#sec-temporal.calendar.prototype.monthCode

I also have problem to understand the reason behind it, it seems both ISOMonth and ISOMonthCode get the information from  [[ISOMonth]] internal slot.

ptomato commented:
@FrankYFTang I created an account on Gerrit and followed up there.

justingrant commented:
Copying @ptomato's note from the other site to close the loop back here: 

> Hi, I can help explain this — the reason is because only the monthCode values remain constant from year to year. The value returned by the month() getter is the ordinal number of the month in a particular year, which may change from year to year in lunisolar calendars.
> 
> For example today's date in the Hebrew calendar is 2 Iyar, as a PlainMonthDay this is monthCode M08 and day 2. In this year, 5782, Iyar is the 9th month because this year has a leap month. Last year, Iyar was the 8th month. Without the year, we can't determine an ordinal value for the month() getter to return.
> 
> (This all doesn't matter for the ISO 8601 calendar which always has 12 months per year, but it would be bad if the month() getter would sometimes return a value and sometimes throw if your code was unexpectedly handed a PlainMonthDay object with a lunisolar calendar.)
